docs/io/fits/index.rst

```diff
@@ -309,6 +309,76 @@ would with a dict::
 
     See also :ref:`sphx_glr_generated_examples_io_modify-fits-header.py`.
 
+.. _structural_keywords:
+
+Structural Keywords
+"""""""""""""""""""
+
+An unfortunate design quirk of FITS is that FITS keywords mix up both metadata
+and critical information file structure that is needed to parse the file.
+These *structural* keywords can be silently modified by :mod:`astropy.io.fits` and,
+in general, should not be touched by the user.
+
+The specific set of structural keywords used by the FITS standard varies with HDU type.
+The following table lists which keywords are associated with each HDU type:
+
+.. csv-table:: Structural Keywords
+   :header: "HDU Type", "Structural Keywords"
+   :widths: 20, 20
+
+   "All", "``SIMPLE``, ``BITPIX``, ``NAXIS``"
+   ":class:`PrimaryHDU`", "``EXTEND``"
+   ":class:`ImageHDU`, :class:`TableHDU`, :class:`BinTableHDU`",  "``PCOUNT``, ``GCOUNT``"
+   ":class:`GroupsHDU`", "``NAXIS1``, ``GCOUNT``, ``PCOUNT``, ``GROUPS``"
+   ":class:`TableHDU`, :class:`BinTableHDU`", "``TFIELDS``, ``TFORM``, ``TBCOL``"
+
+Structural keywords are checked and updated in four different places:
+
+1. When reading a file
+2. When setting the associated data
+3. When setting the :class:`Header` object
+4. When writing a file
+
+This can often cause confusion because any handwritten values users might assign
+to those keywords will be overwrittten. Consider the following examples:
+
+.. code-block:: python
+
+    >>> data = np.random.random((3,3))
+    >>> ph = fits.PrimaryHDU(data)
+
+    # The NAXIS keyword is automatically updated based on the data
+    >>> ph.header['NAXIS']
+    2
+
+    # Changing the data will overwrite any changes the user makes
+    >>> ph.header['NAXIS'] = 0
+    >>> ph.header['NAXIS']
+    0
+    >>> data2 = np.random.random((3,3,3))
+    >>> ph.data = data2
+    >>> ph.header['NAXIS']
+    3
+
+    # Setting the Header object will also trigger a check
+    >>> ph2 = fits.PrimaryHDU()
+    >>> ph2.header['NAXIS']
+    0
+    >>> ph.header['NAXIS'] = 99
+    >>> ph.header = ph2.header
+    >>> ph.header['NAXIS']
+    3
+
+    # Finally, writing a file will trigger a check
+    >>> ph.header['NAXIS'] = 99
+    >>> ph.writeto('tmp.fits')
+    >>> ph.header['NAXIS']
+    3
+
+.. warning::
+  In other words, the structural keywords are a serialization detail that users
+  should almost never manipulate manually except in some rare corner cases.
```

Comment:
I think the "almost never" is enough. It doesn't need the "except in some rare corner cases".